cmd/web-repo-proxy/main_test.go

```diff
@@ -0,0 +1,208 @@
+	repositoriesMutex.Lock()
+	repositories = map[string]*repository{}
+	repositoryDeletionQueue = list.New()
+	repositoriesMutex.Unlock()
+}
+
+func TestMain(m *testing.M) {
+	// Create a temporary directory to store our repositories.
+	repositoriesRoot, _ = ioutil.TempDir("", "web-repo-proxy-test")
+	result := m.Run()
+	os.RemoveAll(repositoriesRoot)
+	os.Exit(result)
+}
+
+type sanitizedCreateRequestTest struct {
+	description         string
+	requestString       string
+	expected            *CreateRepositoryRequest
+	expectedErrorString string
+}
+
+var sanitizedCreateRequestTests = []sanitizedCreateRequestTest{
+	sanitizedCreateRequestTest{
+		description: "Decoding a simple request",
+		requestString: `{
+      "repository_name": "simple_name",
+      "file_contents": {
+        "a_file": "file contents\n"
+      }
+    }`,
+		expected: &CreateRepositoryRequest{
+			RepositoryName: "simple_name",
+			FileContents:   map[string]string{"a_file": "file contents\n"},
+		},
+	},
+	sanitizedCreateRequestTest{
+		description: "Repository name is encoded as a safe url string",
+		requestString: `{
+      "repository_name": "stackoverflow.com/questions/1760757",
+      "file_contents": {
+        "a_file": "file contents\n"
+      }
+    }`,
+		expected: &CreateRepositoryRequest{
+			RepositoryName: "stackoverflow.com%2Fquestions%2F1760757",
+			FileContents:   map[string]string{"a_file": "file contents\n"},
+		},
+	},
+	sanitizedCreateRequestTest{
+		description: "Reject filenames that use subpaths",
+		requestString: `{
+      "repository_name": "simple_name",
+      "file_contents": {
+        "dir/file": "file contents\n"
+      }
+    }`,
+		expectedErrorString: "Repository filenames can't contain '/'",
+	},
+}
+
+func TestSanitizedCreateRequest(t *testing.T) {
+	for _, test := range sanitizedCreateRequestTests {
+		stringReader := strings.NewReader(test.requestString)
+		result, err := sanitizedCreateRequest(stringReader)
+		if !reflect.DeepEqual(result, test.expected) {
+			t.Errorf("[%s] Got %+v, want %+v", test.description, result, test.expected)
+			continue
+		}
+		errorString := ""
+		if err != nil {
+			errorString = err.Error()
+		}
+		if errorString != test.expectedErrorString {
+			t.Errorf("[%s] Expected error [%s], got [%s]", test.description, test.expectedErrorString, errorString)
+		}
+	}
+}
+
+func TestCreateRepository(t *testing.T) {
+	request := httptest.NewRequest("POST", "/create-repository", strings.NewReader(`{
+    "repository_name": "testRepo",
+    "file_contents": {
+      "a_file": "file contents\n"
+    }
+  }`))
+	responseWriter := httptest.NewRecorder()
+	createRepositoryHandler(responseWriter, request)
+
+	result := responseWriter.Result()
+	if result.StatusCode != http.StatusOK {
+		t.Errorf("Create repository returned error code %v", result.StatusCode)
+		return
+	}
+	response := CreateRepositoryResponse{}
+	responseDecoder := json.NewDecoder(result.Body)
+	err := responseDecoder.Decode(&response)
```

Comment:
no need for the intermediate responseDecoder variable

```suggestion
	err := json.NewDecoder(result.Body).Decode(&response)
```